allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+from typing import Callable, Dict, List, Union
+
+import torch
+
+from .common.checks import ConfigurationError
+from .common import Params
+
+
+class NlpApi:
+    """
+    The ``NlpApi`` provides a high-level API for building NLP models.  You don't have to use it
+    for building a model with AllenNLP, but we think you should, because it's useful.
+
+    ``NlpApi`` abstracts away particular decisions that are frequently necessary in NLP, like how
```

Comment:
It's not a magic connection - the dependencies are right there in the constructor.  It's just a bit of re-direction, so the dependencies can be changed later without changing model code.

I don't see a significant difference between what you're suggesting and what we have - your `params.get("my_rnn")` is essentially the same as what we have with `nlp_api.get_recurrent_cell("my_rnn")` - it's just that the "params" are stored inside of `nlp_api` instead of a `params` object (which the model won't have, either, incidentally; it'll just take an `NlpApi` object in its constructor).  Is there some difference that you see that I'm missing?

We're changing the way parameters work in AllenNLP vs. DeepQA - DeepQA made everything take a `Params` object as its only constructor parameter.  Here, we're making dependencies more explicit, where constructors take concrete objects, and just providing a `from_params` method that constructs all necessary dependencies, then calls the constructor.